### PR TITLE
Proj3000: Only use UTF-8 without BOM

### PR DESCRIPTION
**Pull request by Corniel**
There are at least three problems with putting a BOM in UTF-8 encoded files.

1. Files that hold no text are no longer empty because they always contain the BOM.
2. Files that hold text within the ASCII subset of UTF-8 are no longer themselves ASCII because the BOM is not ASCII, which makes some existing tools break down, and it can be impossible for users to replace such legacy tools.
3. It is not possible to concatenate several files together because each file now has a BOM at the beginning.

Source: https://stackoverflow.com/questions/2223882/whats-the-difference-between-utf-8-and-utf-8-with-bom

There are multiple other sources stating basically the same.

### CODE REVIEW

File: src/DotNetProjectFile.Analyzers/Rule.cs [hunk 1]
```diff
@@ -765,6 +765,17 @@ public static class Rule
       tags: ["XML", "indentation"],
       category: Category.Formatting);
 
+    public static DiagnosticDescriptor OnlyUseUTF8WithoutBom => New(
+        id: 3000,
+        title: "Ony use UTF-8 encoding without BOM",
+        message: "This file's encoding using UTF-8 with BOM.",
```
Comment (CptWesley) [inline src/DotNetProjectFile.Analyzers/Rule.cs:771]: this message does not contain a verb

File: src/DotNetProjectFile.Analyzers/Analyzers/Files/OnlyUseUTF8WithoutBom.cs [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+public sealed class OnlyUseUTF8WithoutBom() : ProjectFileAnalyzer<ProjectTextFile>(Rule.OnlyUseUTF8WithoutBom)
+{
+    protected override void Register(AnalysisContext context)
+    {
+        context.RegisterCompilationAction(c =>
+        {
+            if (ProjectFiles.Global.UpdateMsBuildProject(c) is { } msbuild)
+            {
+                foreach (var file in msbuild.Path.Directory.Files("**/*")?
+                    .Where(f => !f.ToString("/").Contains("/bin/"))
```
Comment (CptWesley) [inline src/DotNetProjectFile.Analyzers/Analyzers/Files/OnlyUseUTF8WithoutBom.cs:15]: Just a sanity check, but is is this always `/bin/` on all operating systems?

**Comment by CptWesley** [comment]
Looks good besides my comments

File: src/DotNetProjectFile.Analyzers/Rule.cs (hunk 1 above)
Comment (LauraXiulan) [inline src/DotNetProjectFile.Analyzers/Rule.cs:771]:
```suggestion
        message: "This file's encoding is using UTF-8 with BOM.",
```

File: src/DotNetProjectFile.Analyzers/Analyzers/Files/OnlyUseUTF8WithoutBom.cs (hunk 2 above)
Comment (Corniel) [reply src/DotNetProjectFile.Analyzers/Analyzers/Files/OnlyUseUTF8WithoutBom.cs:15]: See the `.ToString("/")`. This is 'simple' implementation. I'm working on a more robust one.

File: specs/DotNetProjectFile.Analyzers.Specs/Rules/Generic/Only_use_UTF-8_without_BOM.cs [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+
+public class Reports
+{
+    [Test]
+    public void files_with_UTF8_BOM()
+       => new OnlyUseUTF8WithoutBom()
+       .ForProject("MultipleEncodings.cs")
+       .HasIssues(
+           new Issue("Proj3000", "This file's encoding using UTF-8 with BOM.").WithPath("utf8-bom.css"),
+           new Issue("Proj3000", "This file's encoding using UTF-8 with BOM.").WithPath("file.utf8-bom"));
```
Comment (LauraXiulan) [inline specs/DotNetProjectFile.Analyzers.Specs/Rules/Generic/Only_use_UTF-8_without_BOM.cs:13]:
```suggestion
           new Issue("Proj3000", "This file's encoding is using UTF-8 with BOM.").WithPath("utf8-bom.css"),
           new Issue("Proj3000", "This file's encoding is using UTF-8 with BOM.").WithPath("file.utf8-bom"));
```

File: specs/DotNetProjectFile.Analyzers.Specs/Rules/Generic/Only_use_UTF-8_without_BOM.cs [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+}
+
+public class Guards
+{
+    [TestCase("CompliantCSharp.cs")]
```
Comment (LauraXiulan) [inline specs/DotNetProjectFile.Analyzers.Specs/Rules/Generic/Only_use_UTF-8_without_BOM.cs:18]: Can you add an explicit utf8 without bom encoded file to this project

Comment (Corniel) [reply specs/DotNetProjectFile.Analyzers.Specs/Rules/Generic/Only_use_UTF-8_without_BOM.cs:18]: The project files themselves are using UTF-8 without BOM.

**Pull request merged**